gossip/src/received_cache.rs

```diff
@@ -0,0 +1,185 @@
+use {
+    itertools::Itertools,
+    lru::LruCache,
+    solana_sdk::pubkey::Pubkey,
+    std::{cmp::Reverse, collections::HashMap},
+};
+
+// For each origin, tracks which nodes have sent messages from that origin and
+// their respective score in terms of timeliness of delivered messages.
+pub(crate) struct ReceivedCache(LruCache</*origin/owner:*/ Pubkey, ReceivedCacheEntry>);
+
+#[derive(Clone, Default)]
+struct ReceivedCacheEntry {
+    nodes: HashMap<Pubkey, /*score:*/ usize>,
+    num_upserts: usize,
+}
+
+impl ReceivedCache {
+    // Minimum number of upserts before a cache entry can be pruned.
+    const MIN_NUM_UPSERTS: usize = 20;
+
+    pub(crate) fn new(capacity: usize) -> Self {
+        Self(LruCache::new(capacity))
+    }
+
+    pub(crate) fn record(&mut self, origin: Pubkey, node: Pubkey, num_dups: usize) {
+        match self.0.get_mut(&origin) {
+            Some(entry) => entry.record(node, num_dups),
+            None => {
+                let mut entry = ReceivedCacheEntry::default();
+                entry.record(node, num_dups);
+                self.0.put(origin, entry);
+            }
+        }
+    }
+
+    pub(crate) fn prune(
+        &mut self,
+        pubkey: Pubkey, // This node.
+        origin: Pubkey, // CRDS value owner.
```

Comment:
Changed `pubkey` to reference.

The `origin` one is actually tricky because it needs lifetime annotations, and even with those it still breaks the outer `flat_map` here:
https://github.com/solana-labs/solana/blob/637d0358e/gossip/src/crds_gossip_push.rs#L135-L143
due to `referencing data owned by current function`.
Note that the `flat_map` already owns the `origin` so this wouldn't copy unnecessarily.